File: leetcode/2090/solution.py
"""
https://leetcode.com/problems/k-radius-subarray-averages/description/
2090. K Radius Subarray Averages

You are given a 0-indexed array nums of n integers, and an integer k.
The k-radius average for a subarray of nums centered at some index i with the radius k is the average of all elements in nums between the indices i - k and i + k (inclusive). If there are less than k elements before or after the index i, then the k-radius average is -1.
Build and return an array avgs of length n where avgs[i] is the k-radius average for the subarray centered at index i.
The average of x elements is the sum of the x elements divided by x, using integer division. The integer division truncates toward zero, which means losing its fractional part.
For example, the average of four elements 2, 3, 1, and 5 is (2 + 3 + 1 + 5) / 4 = 11 / 4 = 2.75, which truncates to 2.
"""
from typing import List


class Solution:
    def getAverages(self, nums: List[int], k: int) -> List[int]:
        n = k * 2 + 1  # length from i - k to i + k (inclusive)
        avgs = [-1] * len(nums)
        total = sum(nums[: k * 2])

        for i in range(k, len(nums) - k):
            total += nums[i + k]
            avgs[i] = total // n
            total -= nums[i - k]

        return avgs

File: leetcode/2090/solution.rs
impl Solution {
    pub fn get_averages(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let k = k as usize;
        let n = (k * 2 + 1);
        let l = nums.len();
        let mut avgs = vec![-1; l];

        // Edge cage
        if l < k * 2 + 1 {
            return avgs;
        }

        // sum of values from nums[0 to (k * 2 - 1)]
        // e.g. if k =3, then total = nums[0 to 5]
        let mut total: i64 = (0..k * 2).map(|i| nums[i] as i64).sum();

        for i in (k..(l - k)) {
            total += nums[i + k] as i64;
            avgs[i] = (total / (n as i64)) as i32;
            total -= nums[i - k] as i64;
        }

        avgs
    }
}
